pkg/registry/rbac/reconciliation/reconcile_clusterrole.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconciliation
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/rbac"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/rbac/internalversion"
+	"k8s.io/kubernetes/pkg/registry/rbac/validation"
+)
+
+// ReconcileProtectAnnotation is the name of an annotation which prevents reconciliation if set to "true"
+const ReconcileProtectAnnotation = "rbac.authorization.kubernetes.io/reconcile-protect"
+
+type ReconcileOperation string
+
+var (
+	ReconcileCreate   ReconcileOperation = "create"
+	ReconcileUpdate   ReconcileOperation = "update"
+	ReconcileRecreate ReconcileOperation = "recreate"
+	ReconcileNone     ReconcileOperation = "none"
+)
+
+type ReconcileClusterRoleOptions struct {
+	// Role is the expected role that will be reconciled
+	Role *rbac.ClusterRole
+	// Confirm indicates writes should be performed. When false, results are returned as a dry-run.
+	Confirm bool
+	// Union indicates reconciliation should be additive (no permissions should be removed)
+	Union bool
+	// Client is used to look up existing roles, and create/update the role when Confirm=true
+	Client internalversion.ClusterRoleInterface
+}
+
+type ReconcileClusterRoleResult struct {
+	// Role is the resulting role from the reconciliation operation
+	Role *rbac.ClusterRole
+
+	// MissingRules contains expected rules that were missing from the currently persisted role
+	MissingRules []rbac.PolicyRule
+	// ExtraRules contains extra permissions the currently persisted role had
+	ExtraRules []rbac.PolicyRule
+
+	// Operation is the API operation required to reconcile
+	Operation ReconcileOperation
+	// Protected indicates an existing role prevented reconciliation
+	Protected bool
+}
+
+func (o *ReconcileClusterRoleOptions) Run() (*ReconcileClusterRoleResult, error) {
+	var result *ReconcileClusterRoleResult
+
+	existing, err := o.Client.Get(o.Role.Name, metav1.GetOptions{})
+	switch {
+	case errors.IsNotFound(err):
+		result = &ReconcileClusterRoleResult{
```

Comment:
Eh, I can live with it.  Just didn't match my expectations reading top to bottom.